File: src/main/java/org/veupathdb/lib/blast/field/QueryFile.java
package org.veupathdb.lib.blast.field;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.node.TextNode;
import org.veupathdb.lib.blast.util.DefaultingJSONValue;

public record QueryFile(String value) implements DefaultingJSONValue
{
  public static final String DefaultValue = "-";

  @Override
  public boolean isDefault() {
    return value == null || DefaultValue.equals(value);
  }

  @Override
  public JsonNode toJSON() {
    return new TextNode(value);
  }
}

File: src/main/kotlin/org/veupathdb/lib/blast/field/query-file.kt
package org.veupathdb.lib.blast.field

import com.fasterxml.jackson.databind.node.ObjectNode
import org.veupathdb.lib.blast.serial.BlastField


private const val DefaultQueryFile = "-"
private const val KeyQueryFile = "-query"


internal fun ParseQueryFile(js: ObjectNode) =
  QueryFile(js[KeyQueryFile]?.textValue() ?: DefaultQueryFile)


@JvmInline
value class QueryFile(val value: String = DefaultQueryFile) : BlastField {
  override val isDefault get() = value == DefaultQueryFile

  override fun appendJson(js: ObjectNode) {
    if (!isDefault)
      js.put(KeyQueryFile, DefaultQueryFile)
  }

  override fun appendCliSegment(cli: StringBuilder) {
    if (!isDefault)
      cli.append(' ')
        .append(KeyQueryFile)
        .append(" '")
        .append(value)
        .append('\'')
  }

  override fun appendCliParts(cli: MutableList<String>) {
    if (!isDefault) {
      cli.add(KeyQueryFile)
      cli.add(value)
    }
  }
}